File: core/src/main/java/org/strykeforce/thirdcoast/telemetry/util/SignalGenerator.java
package org.strykeforce.thirdcoast.telemetry.util;

/**
 * Generate a signal used for simulation.
 */
public abstract class SignalGenerator {

  // for System.nanoTime() resolution
  private final static double TICKS_PER_SECOND = 1e9;

  protected final double frequency;
  protected final double amplitude;
  protected final double offset;
  protected final double invert;
  protected final double phase;
  private long startTime = System.nanoTime();
  private double time;

  SignalGenerator(double frequency, double phase, double amplitude, double offset,
      double invert) {
    this.frequency = frequency;
    this.phase = phase;
    this.amplitude = amplitude;
    this.offset = offset;
    this.invert = invert;
  }

  abstract double getValue(double time);

  public double getValue() {
    time = (double) (System.nanoTime() - startTime) / TICKS_PER_SECOND;
    return getValue(time);
  }

  public double getTimeForLastValue() {
    return time;
  }

  public void reset() {
    startTime = System.nanoTime();
  }

  public boolean isInverted() {
    return invert == -1.0;
  }

  @Override
  public String toString() {
    return String
        .format("%.2f hz with amplitude %.2f, phase %.2f, offset %.2f%s", frequency, amplitude,
            phase, offset, isInverted() ? ", inverted" : "");

  }

  /**
   * Available signal types.
   */
  public enum SignalType {
    SINE, SQUARE, TRIANGLE, SAWTOOTH
  }

  /**
   * Builder for {@link SignalGenerator}.
   */
  public static class Builder {

    private final SignalType type;
    private double frequency = 1.0;
    private double phase = 0.0;
    private double amplitude = 1.0;
    private double offset = 0.0;
    private double invert = 1.0;

    public Builder(SignalType type) {
      this.type = type;
    }

    public Builder frequency(double val) {
      frequency = val;
      return this;
    }

    public Builder phase(double val) {
      phase = val;
      return this;
    }

    public Builder amplitude(double val) {
      amplitude = val;
      return this;
    }

    public Builder offset(double val) {
      offset = val;
      return this;
    }

    public Builder invert(boolean val) {
      invert = val ? -1.0 : 1.0;
      return this;
    }

    public SignalGenerator build() {
      switch (type) {

        case SINE:
          return new SineSignalGenerator(frequency, phase, amplitude, offset, invert);
        case SQUARE:
          return new SquareSignalGenerator(frequency, phase, amplitude, offset, invert);
        case TRIANGLE:
          return new TriangleSignalGenerator(frequency, phase, amplitude, offset, invert);
        case SAWTOOTH:
          return new SawtoothSignalGenerator(frequency, phase, amplitude, offset, invert);
      }
      return null;
    }
  }
}

File: telemetry/src/main/kotlin/org/strykeforce/thirdcoast/telemetry/util/SignalGenerator.kt
package org.strykeforce.thirdcoast.telemetry.util

private const val TICKS_PER_SECOND = 1e9

/** Generate a signal used for simulation.  */
abstract class SignalGenerator internal constructor(
  protected val frequency: Double,
  protected val phase: Double,
  protected val amplitude: Double,
  protected val offset: Double,
  protected val invert: Double
) {
  private var startTime = System.nanoTime()
  private var timeForLastValue: Double = 0.toDouble()

  val value: Double
    get() {
      timeForLastValue = (System.nanoTime() - startTime).toDouble() / TICKS_PER_SECOND
      return getValue(timeForLastValue)
    }

  private val isInverted: Boolean
    get() = invert == -1.0

  internal abstract fun getValue(time: Double): Double

  fun reset() {
    startTime = System.nanoTime()
  }

  override fun toString(): String {
    return String.format(
      "%.2f hz with amplitude %.2f, phase %.2f, offset %.2f%s",
      frequency, amplitude, phase, offset, if (isInverted) ", inverted" else ""
    )
  }

  /** Available signal types.  */
  enum class SignalType {
    SINE,
    SQUARE,
    TRIANGLE,
    SAWTOOTH
  }

  /** Builder for [SignalGenerator].  */
  class Builder(private val type: SignalType) {
    private var frequency = 1.0
    private var phase = 0.0
    private var amplitude = 1.0
    private var offset = 0.0
    private var invert = 1.0

    fun frequency(`val`: Double): Builder {
      frequency = `val`
      return this
    }

    fun phase(`val`: Double): Builder {
      phase = `val`
      return this
    }

    fun amplitude(`val`: Double): Builder {
      amplitude = `val`
      return this
    }

    fun offset(`val`: Double): Builder {
      offset = `val`
      return this
    }

    fun invert(`val`: Boolean): Builder {
      invert = if (`val`) -1.0 else 1.0
      return this
    }

    fun build(): SignalGenerator = when (type) {
      SignalGenerator.SignalType.SINE -> SineSignalGenerator(
        frequency,
        phase,
        amplitude,
        offset,
        invert
      )
      SignalGenerator.SignalType.SQUARE -> SquareSignalGenerator(
        frequency,
        phase,
        amplitude,
        offset,
        invert
      )
      SignalGenerator.SignalType.TRIANGLE -> TriangleSignalGenerator(
        frequency,
        phase,
        amplitude,
        offset,
        invert
      )
      SignalGenerator.SignalType.SAWTOOTH -> SawtoothSignalGenerator(
        frequency,
        phase,
        amplitude,
        offset,
        invert
      )
    }
  }

}
